kotlin/iwa-sample-app/src/main/java/com/esri/android/samples/iwasampleapp/MainActivity.kt

```diff
@@ -0,0 +1,173 @@
+ * 
+ * When a user credential challenge is issues a dialog will be presented for credential.
+ * The portal url will be displayed as message in the dialog. If wrong credential has been
+ * passed in the previous attempt, a different message will be display in the dialog.
+ * The dialog has two edit text boxes for username and password respectively. Other SDKs'
+ * sample may have one more parameter for IWA domain. As indicated by the Javadoc of UseCredential
+ * our SDK is in favor of passing username as username@domain. This sample doesn't provide
+ * a edit text box for domain to simplify the logic.
+ *  
+ * @see <a href="https://developers.arcgis.com/android/latest/api-reference/reference/com/esri/arcgisruntime/security/UserCredential.html#UserCredential(java.lang.String,%20java.lang.String)">Javadoc of UserCredential</a>
+ * 
+ */
+class IWACustomChallengeHandler(val activity: Activity) : AuthenticationChallengeHandler {
+    override fun handleChallenge(challenge: AuthenticationChallenge): AuthenticationChallengeResponse {
+        if (challenge.type == AuthenticationChallenge.Type.USER_CREDENTIAL_CHALLENGE) {
+            if (challenge.failureCount > MAX_ATTEMPTS) {
+                // Exceed maximum amount of attempts. Act like it was a cancel
+                Toast.makeText(activity, "Exceed maximum amount of attempts. Please try again!", Toast.LENGTH_SHORT).show()
+                return AuthenticationChallengeResponse(AuthenticationChallengeResponse.Action.CANCEL, challenge)
+            }
+            
+            // Create a countdown latch with a count of one to synchronize the dialog
+            val signal : CountDownLatch = CountDownLatch(1)
+            var credential : UserCredential?
+            credential = null
+            
+            // Present the sign-in dialog
+            activity.runOnUiThread(java.lang.Runnable { 
+                // Inflate the layout
+                val dialogView = activity.layoutInflater.inflate(R.layout.credential_dialog, null)
+                // Create the dialog
+                val builder: AlertDialog.Builder = AlertDialog.Builder(activity)
+                val hostname = dialogView.auth_hostname
+                val username = dialogView.auth_username
+                val password = dialogView.auth_password
+                // Set click listeners
+                builder.setPositiveButton("Sign In") {dialog, which ->
+                            // Create user credential
+                            credential = UserCredential(username.text.toString(), password.text.toString())
+                            signal.countDown()
+                        }
+                        .setNegativeButton("Cancel") {dialog, which ->
+                            // User cancel the loading process
+                            val remoteResource = challenge.remoteResource
+                            if (remoteResource is Loadable) {
+                                remoteResource.cancelLoad()
```

Comment:
I don't understand why we explicitly cancel the loading of the RemoteResource here. I assume that's the Portal instance? It should just fail to load implicitly if we don't get any credentials.

Reply:
updated

Reply:
I reverted this specific change. One of the purposes of this sample is to show a workaround for a known issue reported by Indoors team, which is when users hit the “Cancel” button in the dialog prompting for username/password or the “Back” button on the phone , the SDK throws the same error as that user is not authenticated. The SDK should distinguish these different scenarios and provide different error code/message. As a workaround, I explicitly cancel the loading of the RemoteResource as the operation when the user tapping on the "Cancel" button.

>  I assume that's the Portal instance?

Yes, in this sample it is the Portal instance. But this custom challenge handler should work with other RemoteResource types. I use RemoteResource to make it more general.